### PR TITLE
Feature: #18 include buildtemplate in createdeltapkg + #300 no more key-name swapping in rt

### LINKED ISSUE #18: include buildTemplate in createDeltaPkg instead of retrieveAsTemplate / CSCLSROZ-555
the current approach does not actually use the files committed to git but retrieves again from the server -that might break things given that DEV might have already changed by the time this runs.

therefore, buildTemplate must be added which does what retrieveAsTemplate does, only completely local, without a retrieve

### LINKED ISSUE #300: [BREAKING CHANGE] Disable key/name override during retrieveAsTemplate
this is malfunctioning if key & name are not equal and is overall causing confusion.

### PR DESCRIPTION
**Pull request by JoernBerkefeld**
# PR details

## What is the purpose of this pull request? (put an "X" next to an item)

_Please delete options that are not relevant._

- [ ] Documentation update
- [ ] Bug fix
- [ ] New metadata support
- [ ] Enhanced metadata
- [ ] Add a CLI option
- [x] Add something to the core
- [ ] Other, please explain:

## What changes did you make? (Give an overview)

- closes #300 (breaking change)
- closes #18 
-

## Is there anything you'd like reviewers to focus on?

**please merge PR #299 first.**

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ESLint & Prettier are not outputting errors or warnings
- [ ] README.md updated (if applicable)

### CODE REVIEW

**Comment by JoernBerkefeld** [comment]
@DougMidgley this one is ready for review now

File: lib/Builder.js [hunk 1]
```diff
@@ -178,6 +294,28 @@ class Builder {
             }
         }
     }
+    /**
+     * check if a market name exists in current mcdev config
+     *
+     * @param {string} market market localizations
+     * @returns {boolean} found market or not
+     */
+    static _checkMarket(market) {
```
Comment (DougMidgley) [inline lib/Builder.js:303]: this feels a lot like it should be in util not in builder?

File: lib/Builder.js [hunk 2]
```diff
@@ -178,6 +294,28 @@ class Builder {
...
+        const properties = File.loadConfigFile();
+        if (properties.markets[market]) {
+            return true;
+        } else {
+            Util.logger.error(`Could not find the market '${market}' in your configuration file.`);
+            const marketArr = [];
+            for (const oneMarket in properties.markets) {
```
Comment (DougMidgley) [inline lib/Builder.js:310]: const martketArr = Object.values(properties.markets) would be preferred

File: lib/util/devops.js [hunk 3]
```diff
@@ -186,7 +191,7 @@ const DevOps = {
         }
         // Write into delta.json to serve as documentation
         File.writeJSONToFile(properties.directories.deltaPackage, 'delta_package', delta);
-        this.document(properties.directories.deltaPackage, delta);
+        DevOps.document(properties.directories.deltaPackage, delta);
```
Comment (DougMidgley) [inline lib/util/devops.js:194]: any reason to change this?

File: lib/util/devops.js [hunk 4]
```diff
@@ -237,6 +243,178 @@ const DevOps = {
...
+                            name: 'isPurgeDeployFolder',
+                            message:
+                                'Do you want to empty the deploy folder (ensures no files from previous deployments remain)?',
+                            default: true,
+                        },
+                    ]);
+                }
+                if (skipInteraction || responses.isPurgeDeployFolders) {
+                    // Clear output folder structure for selected sub-type
+                    File.removeSync(File.normalizePath([properties.directories.deploy]));
```
Comment (DougMidgley) [inline lib/util/devops.js:389]: prefer async

File: lib/util/devops.js [hunk 5]
```diff
@@ -237,6 +243,178 @@ const DevOps = {
...
         return delta;
     },
+    /**
+     * wrapper around DevOps.getDeltaList, Builder.buildTemplate and M
+     *
+     * @param {object} properties project config file
+     * @param {string} range git commit range
+     * @param {TYPE.SkipInteraction} [skipInteraction] allows to skip interactive wizard
+     */
+    async buildDeltaDefinitions(properties, range, skipInteraction) {
```
Comment (DougMidgley) [inline lib/util/devops.js:253]: part of me feels like this may benefit from being broken down into a couple of diff methods.

**Review by DougMidgley — CHANGES_REQUESTED** [review]

File: lib/util/devops.js (hunk 3 above)
Comment (JoernBerkefeld) [reply lib/util/devops.js:194]: nope, not really. consistency maybe

File: lib/util/devops.js (hunk 5 above)
Comment (JoernBerkefeld) [reply lib/util/devops.js:253]: i agree. not really in scope here though but let me create a ticket.

Comment (JoernBerkefeld) [reply lib/util/devops.js:253]: created #318

File: lib/Builder.js (hunk 1 above)
Comment (DougMidgley) [reply lib/Builder.js:303]: yes, but in the utils file, you could avoid getting loadConfigFile, and just pass the properties in. I looked at all references and all have properties object available before using this method. I can imagine it being useful for other items in future elsewhere :/ not about to fall on my sword about it given its only open item - we could reopen in a new point about removing this cyclical dependency item

**Review by DougMidgley — CHANGES_REQUESTED** [review]

File: lib/Builder.js (hunk 1 above)
Comment (JoernBerkefeld) [reply lib/Builder.js]: done

**Review by DougMidgley — APPROVED** [review]

**Pull request merged**